lib/AST/Attr.cpp

```diff
@@ -440,9 +441,39 @@ bool DeclAttribute::printImpl(ASTPrinter &Printer, const PrintOptions &Options)
   case DAK_Specialize: {
     Printer << "@" << getAttrName() << "(";
     auto *attr = cast<SpecializeAttr>(this);
-    interleave(attr->getTypeLocs(),
-               [&](TypeLoc tyLoc){ tyLoc.getType().print(Printer, Options); },
-               [&]{ Printer << ", "; });
+    auto exported = attr->isExported() ? "true" : "false";
+    auto kind = attr->isPartialSpecialization() ? "partial" : "full";
+
+    Printer << "exported: "<<  exported << ", ";
+    Printer << "kind: " << kind << ", ";
+
+    if (!attr->getRequirements().empty()) {
+      Printer << "where ";
+    }
+    std::function<Type(Type)> GetInterfaceType;
+    auto *FnDecl = attr->getDecl();
+    if (!FnDecl)
+      GetInterfaceType = [](Type Ty) -> Type { return Ty; };
+    else {
+      // Use GenericEnvironment to produce user-friendly
+      // names instead of something like t_0_0.
+      auto *GenericEnv = FnDecl->getGenericEnvironment();
+      auto *SM = FnDecl->getModuleContext();
+      assert(GenericEnv);
+      GetInterfaceType = [=](Type Ty) -> Type {
+        return GenericEnv->mapTypeOutOfContext(
```

Comment:
@slavapestov getSugaredType takes only `GenericTypeParamType` as its parameters. It is impossible to call it with a Type (which may be something containing GenericTypeParamType). So, one would need a `type.walk` here. It is not clear to me if it would better that this MapInto/MapOutOf Context trick.